File: Baekjoon/완전탐색(Brute Force)/10974 모든 순열/HwangInGyu.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int N = Integer.parseInt(st.nextToken());
        boolean[] isVisited = new boolean[N];

        int idxOfCurrentNode = 0;
        int[] answer = new int[N];
        recursion(idxOfCurrentNode, isVisited, answer);
        System.out.println(sb);
    }

    private static void recursion(int idxOfCurrentNode, boolean[] isVisited, int[] answer) {
        if (idxOfCurrentNode == isVisited.length) {
            for (int i : answer) {
                sb.append(i);
                sb.append(" ");
            }
            sb.append("\n");
        }

        for (int i = 0; i < answer.length; i++) {
            if (isVisited[i]) {
                continue;
            }
            isVisited[i] = true;
            answer[idxOfCurrentNode] = i+1;
            recursion(idxOfCurrentNode+1, isVisited, answer);
            isVisited[i] = false;
            answer[idxOfCurrentNode] = 0;
        }
    }
}

File: Baekjoon/Graph Theory/1260 DFS와 BFS/HwangInGyu.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (cntOfNode, cntOfDegree, positionOfStartNode) = readLine().split(" ").map { it -> it.toInt() }
    val graph = Array<LinkedList<Int>>(cntOfNode) {LinkedList()}

    for (noOfDegree in 0 until cntOfDegree) {
        val (startNode, endNode) = readLine().split(" ").map { it.toInt()-1 }
        graph[startNode].add(endNode)
        graph[endNode].add(startNode)
    }

    graph.map { it.sort() }
    var isVisited = BooleanArray(cntOfNode)
    println(dfs(StringBuilder(), positionOfStartNode-1, graph, isVisited))
    isVisited = BooleanArray(cntOfNode)
    println(bfs(positionOfStartNode-1, graph, isVisited))
}

fun dfs(
    answer: StringBuilder,
    positionOfStartNode: Int,
    graph: Array<LinkedList<Int>>,
    visited: BooleanArray
): StringBuilder {
    if (visited[positionOfStartNode]) {
        return answer
    }

    visited[positionOfStartNode] = true
    answer.append(positionOfStartNode+1).append(" ")
    graph[positionOfStartNode].forEach { nextPosition ->
        answer.append(dfs(StringBuilder(), nextPosition, graph, visited))
    }

    return answer
}

fun bfs(
    positionOfStartNode: Int,
    graph: Array<LinkedList<Int>>,
    visited: BooleanArray
): StringBuilder {
    val sb = StringBuilder()
    val queue = LinkedList<Int>()
    sb.append(positionOfStartNode+1).append(" ")
    visited[positionOfStartNode] = true

    queue.offer(positionOfStartNode)
    while (!queue.isEmpty()) {
        val currentNode = queue.poll()

        graph[currentNode].forEach { nextNode ->
            if (!visited[nextNode]) {
                queue.offer(nextNode)
                sb.append(nextNode+1).append(" ")
                visited[nextNode] = true
            }
        }
    }

    return sb
}